elastic/kibana
# [Alerting] Initial implementation of alerting task `cancel()`

Resolves https://github.com/elastic/kibana/issues/113459

## Summary

With this PR, the alerting task runner implements a `cancel()` function (currently only called by the task manager when an alerting task exceeds its timeout value).

* writes an Event Log document with action `execute-timeout`
* updates the execution status in the rule saved object to reflect the timeout with `status:error`, `reason:timeout`.
* adds alerting config `cancelAlertsOnRuleTimeout` to control action skipping behavior at framework level, defaults to `true`
* adds rule type config `cancelAlertsOnRuleTimeout` to control action skipping behavior at rule type level. if not set, defaults to whatever the value of the alerting config `cancelAlertsOnRuleTimeout` is
* skips scheduling actions if the execution is cancelled AND the `cancelAlertsOnRuleTimeout` in `kibana.yml` is true AND the rule type `cancelAlertsOnRuleTimeout` is true

## To Verify

1. Modify the Elasticsearch query rule type `x-pack/plugins/stack_alerts/server/alert_types/es_query/alert_type.ts`:

```
@@ -148,6 +148,7 @@ export function getAlertType(logger: Logger): AlertType<
     isExportable: true,
     executor,
     producer: STACK_ALERTS_FEATURE_ID,
+    ruleTaskTimeout: '1m',
   };

   async function executor(
@@ -170,6 +171,8 @@ export function getAlertType(logger: Logger): AlertType<
       throw new Error(getInvalidComparatorError(params.thresholdComparator));
     }

+    await new Promise((resolve) => setTimeout(resolve, 120000));
```

This sets the `ruleTaskTimeout` to `1m` and adds a 2 minute delay to the executor

2. Create an Elasticsearch query rule (make sure it is one that will fire!) with an attached action and let it run.
3. After 1 minute, you should see this warning in the logs:

```
[WARN ][plugins.taskManager] Cancelling task alerting:.es-query "f15847e0-2a8e-11ec-b690-17b293411601" as it expired at 2021-10-11T12:30:58.026Z after running for 01m 02s (with timeout set at 1m).
```

4. In the rule management view, the rule should be in an error state
<img width="1243" alt="Screen Shot 2021-10-11 at 8 35 50 AM" src="https://user-images.githubusercontent.com/13104637/136790845-54957fc2-ba4f-4cc1-9a38-e1b3fbe92e37.png">

5. Verify in the kibana index that the rule SO execution status has `error` status with `reason: timeout`:

<details>
<summary> Rule SO execution status </summary>

```
{
  "executionStatus" : {
    "status" : "error",
    "lastExecutionDate" : "2021-10-11T12:36:01.073Z",
    "error" : {
      "reason" : "timeout",
      "message" : "rule execution cancelled due to timeout: \".es-queryf08d2fb0-2a8e-11ec-b690-17b293411601\""
    }
  }
}
```

</details>

6. Verify in the event log index there is a document for `execute-timeout`:

<details>
<summary> Event log doc </summary>

```
{
	"@timestamp": "2021-10-11T12:31:00.923Z",
	"event": {
		"provider": "alerting",
		"action": "execute-timeout",
		"kind": "alert",
		"category": [
			"stackAlerts"
		]
	},
	"message": "rule execution cancelled due to timeout: \".es-queryf08d2fb0-2a8e-11ec-b690-17b293411601\"",
	"kibana": {
		"saved_objects": [{
			"rel": "primary",
			"type": "alert",
			"id": "f08d2fb0-2a8e-11ec-b690-17b293411601",
			"type_id": ".es-query"
		}],
		"server_uuid": "5b2de169-2785-441b-ae8c-186a1936b17d",
		"version": "8.0.0"
	},
	"rule": {
		"id": "f08d2fb0-2a8e-11ec-b690-17b293411601",
		"license": "basic",
		"category": ".es-query",
		"ruleset": "stackAlerts"
	},
	"ecs": {
		"version": "1.8.0"
	}
}
```

</details>

7. Verify you receive no action

8. Now add `xpack.alerting.cancelAlertsOnRuleTimeout: false` to your kibana config and repeat steps 1-7. Verify that all the timeout docs/statuses get written as expected, but you also receive your action at the end of rule execution (after timeout).

9. Now remove `xpack.alerting.cancelAlertsOnRuleTimeout: false` from your kibana config and add `cancelAlertsOnRuleTimeout: false` to the Elasticsearch rule type definition in `x-pack/plugins/stack_alerts/server/alert_types/es_query/alert_type.ts`. Repeat steps 1-7 and verify that all the timeout docs/statuses get written as expected, but you also receive your action at the end of rule execution (after timeout).


## Checklist

Delete any items that are not applicable to this PR.
